File: Median_of_Two_Sorted_Arrays.js
/**Given two sorted arrays nums1 and nums2 of size m and n respectively, return the median of the two sorted arrays.

The overall run time complexity should be O(log (m+n)).**/

function findMedianSortedArrays(nums1, nums2) {
  const merged = mergeSortedArrays(nums1, nums2);
  const len = merged.length;
  const mid = Math.floor(len / 2);

  if (len % 2 === 0) {
    return (merged[mid] + merged[mid - 1]) / 2;
  } else {
    return merged[mid];
  }
}

function mergeSortedArrays(nums1, nums2) {
  const merged = [];
  let i = 0;
  let j = 0;

  while (i < nums1.length && j < nums2.length) {
    if (nums1[i] <= nums2[j]) {
      merged.push(nums1[i]);
      i++;
    } else {
      merged.push(nums2[j]);
      j++;
    }
  }

  while (i < nums1.length) {
    merged.push(nums1[i]);
    i++;
  }

  while (j < nums2.length) {
    merged.push(nums2[j]);
    j++;
  }

  return merged;
}

console.log(findMedianSortedArrays([1,2], [3,4]))//Output: 2.5

File: Median_of_Two_Sorted_Arrays.py
#Given two sorted arrays nums1 and nums2 of size m and n respectively, return the median of the two sorted arrays.
#The overall run time complexity should be O(log (m+n)).

def findMedianSortedArrays(nums1, nums2):
    merged = mergeSortedArrays(nums1, nums2)
    length = len(merged)
    mid = length // 2

    if length % 2 == 0:
        return (merged[mid] + merged[mid - 1]) / 2
    else:
        return merged[mid]

def mergeSortedArrays(nums1, nums2):
    merged = []
    i = 0
    j = 0

    while i < len(nums1) and j < len(nums2):
        if nums1[i] <= nums2[j]:
            merged.append(nums1[i])
            i += 1
        else:
            merged.append(nums2[j])
            j += 1

    while i < len(nums1):
        merged.append(nums1[i])
        i += 1

    while j < len(nums2):
        merged.append(nums2[j])
        j += 1

    return merged

print(findMedianSortedArrays([1,3], [2]))
